### PR TITLE
Add diagnostic loggable properties to neighbor lists.

### LINKED ISSUE #1265: Add properties to NeighborList
## Description

Add properties such as `builds` to NeighborList. See the `printStats` method in HOOMD v2 for a summary of what properties might be of interest to users.

## Motivation and context

In HOOMD v2, NeighborList would print useful stats (like the number of builds) at the end of every call to `run`. This has been removed in v3, but users may still want to access that information.

### PR DESCRIPTION
**Pull request by joaander**
## Description

- Add `num_builds` to `hoomd.md.nlist.NeibhborList`.
- Add `dimensions` to `hoomd.md.nlist.Cell`.
- Add `allocated_particles_per_cell` to `hoomd.md.nlist.Cell`.

## Motivation and context

When testing and debugging performance and memory usage, it is useful to know how many neighbor list builds are performed and house much memory the cell list uses. HOOMD v2 used to print this information at the end of every run, but that was removed in v3.

v2 also used to provide statistics on the min/max/average neighbors per particle and particles per cell. This takes significant extra time to compute and is not as useful when testing.

Resolves #1265

## How has this been tested?

I added unit tests.

## Change log

```
*Added*:

- ``hoomd.md.nlist.Neighborlist.num_builds`` property - The number of neighbor list builds since the last call to ``Simulation.run``.
- ``hoomd.md.nlist.Cell.dimensions`` property - The dimensions of the cell list.
- ``hoomd.md.nlist.Cell.allocated_particles_per_cell`` property -  The number of particle slots allocated per cell.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/md/pytest/test_nlist.py [hunk 1]
```diff
@@ -145,10 +145,60 @@ def test_pickling(simulation_factory, two_particle_snapshot_factory):
...
+
+    sim = simulation_factory(lattice_snapshot_factory(n=10))
+    sim.operations.integrator = integrator
+
+    sim.run(10)
+
+    assert nlist.num_builds == 10
+    assert nlist.shortest_rebuild == 1
+    dim = nlist._cpp_obj.getDim()
+    dim = nlist.dimensions
```
Comment (AlainKadar) [inline hoomd/md/pytest/test_nlist.py]: Why define this twice?

Comment (joaander) [reply hoomd/md/pytest/test_nlist.py]: Good catch. I was finding the source of an `AttributeError`. Due to the way that Python processes `AttributeErrors`, the 2nd line made it hard to find the cause of the error, so I added the first. The first shouldn't be there, it was just for debugging.

**Review by AlainKadar — COMMENTED** [review]
I have just added one comment on something I don't understand.

File: hoomd/md/pytest/test_nlist.py (hunk 1 above)
Comment (AlainKadar) [reply hoomd/md/pytest/test_nlist.py]: Ok, thanks

File: hoomd/md/nlist.py [hunk 2]
```diff
@@ -202,6 +211,28 @@ def _attach(self):
...
+    @log(requires_run=True, default=False, category='sequence')
+    def dimensions(self):
+        """tuple[int, int, int]: Cell list dimensions.
+
+        `dimensions` is the number of cells in the x, y, and z directions.
+
+        See Also:
+            `allocated_particles_per_cell`
+        """
+        print(dir(self._cpp_obj))
```
Comment (b-butler) [inline hoomd/md/nlist.py]: remove
```suggestion
```

File: hoomd/md/pytest/test_nlist.py [hunk 3]
```diff
@@ -145,10 +145,59 @@ def test_pickling(simulation_factory, two_particle_snapshot_factory):
...
+            },
+            'dimensions': {
+                'category': LoggerCategories.sequence,
+                'default': False
+            },
+            'allocated_particles_per_cell': {
+                'category': LoggerCategories.scalar,
+                'default': False
+            },
+        })
```
Comment (b-butler) [inline hoomd/md/pytest/test_nlist.py]:
```suggestion
    base_loggables =  {
        'shortest_rebuild': {'category': LoggerCategories.scalar,  'default': True},
        'num_builds': {'category': LoggerCategories.scalar, 'default': False}
    }
    logging_check(
        hoomd.md.nlist.NeighborList, ('md', 'nlist'), base_loggables)

    logging_check(
        hoomd.md.nlist.Cell, ('md', 'nlist'), {**base_loggables,
            'dimensions': {
                'category': LoggerCategories.sequence,
                'default': False
            },
            'allocated_particles_per_cell': {
                'category': LoggerCategories.scalar,
                'default': False
            },
        })
```

**Review by b-butler — CHANGES_REQUESTED** [review]
Looks good just a couple small changes.

**Review by b-butler — APPROVED** [review]

**Pull request merged**